README.md
# Rack::SteadyETag

`Rack::SteadyETag` is a Rack middleware that generates the same default [`ETag`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) for responses that only differ in XOR-masked CSRF tokens or CSP nonces.

By default Rails uses [`Rack::ETag`](https://rdoc.info/github/rack/rack/Rack/ETag) to generate `ETag` headers by hashing the response body. In theory this would [enable caching](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) for multiple requests to the same resource. However, since most Rails application layouts insert randomly rotating CSRF tokens and CSP nonces into the HTML, two requests for the same content and user will never produce the same response bytes. This means `Rack::ETag` will never send the same ETag twice, causing responses to [never hit a cache](https://github.com/rails/rails/issues/29889).

`Rack::SteadyETag` is a drop-in replacement for `Rack::ETag`. It excludes random content (like CSRF tokens) from the generated ETag, causing two requests for the same content to usually carry the same ETag.

## What is ignored

`Rack::SteadyTag`  ignores the following patterns from the `ETag` hash:

```html
<meta name="csrf-token" value="random" ...>
<meta name="csp-nonce" value="random" ...>
<input name="authenticity_token" value="random" ...>
<script nonce="random" ...> <!-- only the [nonce] attribute -->
```

You can add your own patterns:

```ruby
Rack::SteadyETag::STRIP_PATTERNS << /<meta name="XSRF-TOKEN" value="[^"]+">/
```

You can also push lambda for arbitrary transformations:

```ruby
Rack::SteadyETag::STRIP_PATTERNS << -> { |text| text.gsub(/<meta name="XSRF-TOKEN" value="[^"]+">/, '') }
```

Transformations are only applied for the `ETag` hash. The response body will not be changed.

## What responses are processed

This middleware will process responses that match all of the following: 

- Responses with a HTTP status of 200 or 201.
- Responses with a `Content-Type` of `text/html` or `application/xhtml+xml`.
- Responses with a body. 

Responses should also have an UTF-8 encoding (not checked by the middleware).

This middleware can also add a default `Cache-Control` header for responses it *didn't* process. This is passed as an argument during middleware initialization (see *Installation* below). 

## Covered edge cases

- Different `ETags` are generated when the same content is accessed with different Rack sessions.
- Different `ETags` are generated when a Rails controller manually rotates the CSRF token.
- `ETags` are only generated when the response is `Cache-Control: private` (this is a default in Rails).
- No `ETag` is generated when the response already has an `ETag` header.
- No `ETag` is generated when the response already has an `Last-Modified` header.


## Installation in Rails

Add this line to your application's Gemfile:

```ruby
gem 'rack-steady_etag'
```

And then execute:

```bash
bundle install
```

Make an initializer `config/initializer/etags.rb`:

```ruby
Rails.application.config.middleware.swap Rack::ETag, Rack::SteadyETag, 'no-cache'
```

The `'no-cache'` argument is the default `Cache-Control` for responses that cannot be digested. While it may feel surprising that the middleware changes the `Cache-Control` header in such a case, the [Rails default middleware stack](https://github.com/rails/rails/blob/d96609505511a76c618dc3adfa3ca4679317d008/railties/lib/rails/application/default_middleware_stack.rb#L81) configures the same behavior.


## Development

- After checking out the repo, run `bin/setup` to install dependencies.
- Run `bundle exec rspec` to run the tests.
- Run `BUNDLE_GEMFILE=Gemfile.rack1 bundle exec rspec` to run tests for old Rack 1.
- You can also run `bin/console` for an interactive prompt that will allow you to experiment.
- To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Credits

This library is based on `Rack::ETag`, created by the [Rack Core Team](https://github.com/rack/rack#label-Thanks) and [Rack contributors](https://github.com/rack/rack/graphs/contributors).

Additional changes by [Henning Koch](https://twitter.com/triskweline) from [makandra](https://makandra.com).

## Limitations

- No streaming support. [This will be broken until at least Rack 3](https://github.com/rack/rack/issues/1619). This is not a use case of mine.


CHANGELOG.md
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

- Run local tests against Ruby 3.2 instead of 3.0 from now on




## 0.3.2 - 2022-07-29

- Digest includes the [unmasked Rails CSRF token](https://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection.html#method-i-real_csrf_token) in case a Rails controller manually rotates the token.  

## 0.3.1 - 2022-07-19

- Fix a bug where we would not strip HTML responses with an embedded charset (e.g. `text/html; charset=utf-8`).

## 0.3.0 - 2022-05-13

- Support for old Rack 1.4.7 (last version supported by Rails 3.2)
- No longer depends on activesupport.

## 0.2.3 - 2022-05-12

- Don't depend on `byebug` being in the user bundle.

## 0.2.2 - 2022-05-12

- Don't raise an error when processing binary content.

## 0.2.1 - 2022-05-12

- Only strip patterns for HTML and XHTML responses.

## 0.2.0 - 2022-05-12

- Be more compatible with Rack 2.2.3:
  - Always set a `Cache-Control` header, even for responses that we don't try to digest.
- Strip patterns for responses with `Cache-Control: public`
- Requires Rack 2.x (we want to break with Rack 3)

## 0.1.1 - 2022-05-16

- Activate rubygems MFA

## 0.1.0 - 2021-12-01

- initial release


spec/spec_helper.rb
# frozen_string_literal: true

require "rack/steady_etag"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/rack/steady_etag_spec.rb
# Based on tests for Rack::Etag
# https://github.com/rack/rack/blob/master/test/spec_etag.rb

require 'rack/session/abstract/id'
require 'rack/mock'
require 'rack/lint'

describe Rack::SteadyEtag do

  def etag(app, *args, **kwargs)
    Rack::Lint.new Rack::SteadyETag.new(app, *args, **kwargs)
  end

  def request(opts = {})
    Rack::MockRequest.env_for("", opts)
  end

  def session(id:, **other_props)
    session = { 'session_id' => id }

    # Session keys are strings. Since we don't have ActiveSupport,
    # we manually stringify keys.
    other_props.each do |key, value|
      session[key.to_s] = value
    end

    session
  end

  def sendfile_body
    File.new(File::NULL)
  end

  def text_response(html, headers: {}, env: {})
    app = lambda { |env| [200, headers, [html]] }
    etag(app).call(request(env))
  end

  def html_response(html, headers: {}, env: {})
    default_headers = { 'Content-Type' => 'text/html' }
    headers = default_headers.merge(headers)
    text_response(html, headers: headers, env: env)
  end

  matcher :have_same_etag_as do |response2|
    match do |response1|
      etag1 = response1[1]['ETag']
      etag2 = response2[1]['ETag']

      !etag1.nil? && etag1 != '' && !etag2.nil? && etag2 != '' && etag1 == etag2
    end

    failure_message do |response1|
      "expected '#{etag(response1)}' to be the same ETag as #{etag_from_response(response2)}"
    end

    failure_message_when_negated do |response1|
      "expected responses to have different ETags, but both had #{etag_from_response(response1)}"
    end

    def etag_from_response(response)
      response[1]['ETag']
    end
  end

  matcher :have_body do |expected_string|
    match do |actual_body|
      actual_string = ''
      actual_body.each do |part|
        actual_string << part
      end

      actual_string == expected_string
    end
  end

  it 'generates the same ETag for two equal response bodies' do
    response1 = html_response('Foo')
    response2 = html_response('Foo')

    expect(response1).to have_same_etag_as(response2)
  end

  it 'generates different ETags for different response bodies' do
    response1 = html_response('Foo')
    response2 = html_response('Bar')

    expect(response1).not_to have_same_etag_as(response2)
  end

  it "generates the same ETags for two bodies that only differ in a <meta name='csrf-token'>" do
    response1 = html_response(<<~HTML)
      <head>
        <meta name="csrf-token" content="6EueAlhls9P" />
      </head>
    HTML

    response2 = html_response(<<~HTML)
      <head>
        <meta name="csrf-token" content="qMN0fkVqOg" />
      </head>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it "ignores attribute order in an <meta name='csrf-token'>" do
    response1 = html_response(<<~HTML)
      <head>
        <meta content="6EueAlhls9P" name="csrf-token" />
      </head>
    HTML

    response2 = html_response(<<~HTML)
      <head>
        <meta content="qMN0fkVqOg" name="csrf-token" />
      </head>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it 'does not change the response body when ignoring content' do
    html = <<~HTML
       <head>
        <meta name="csrf-token" content="6EueAlhls9P" />
      </head>
    HTML

    response = html_response(html.dup)
    expect(response[2]).to have_body(html.dup)
  end

  it "generates the same ETags for two bodies that only differ in form's authenticity token token" do
    response1 = html_response(<<~HTML)
      <form>
        <input type="hidden" name="authenticity_token" content="123" />
      </form>
    HTML

    response2 = html_response(<<~HTML)
      <form>
        <input type="hidden" name="authenticity_token" content="456" />
      </form>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it "generates the same ETags for two bodies that only differ in a <meta name='csp-nonce'>" do
    response1 = html_response(<<~HTML)
      <head>
        <meta name="csp-nonce" content="123" />
      </head>
    HTML

    response2 = html_response(<<~HTML)
      <head>
        <meta name="csrf-token" content="456" />
      </head>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it "ignores attribute order in an <meta name='csp-nonce'>" do
    response1 = html_response(<<~HTML)
      <head>
        <meta content="123" name="csp-nonce" />
      </head>
    HTML

    response2 = html_response(<<~HTML)
      <head>
        <meta content="456" name="csrf-token" />
      </head>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it "generates the same ETags for two bodies that only differ in a <script nonce>" do
    response1 = html_response(<<~HTML)
      <script nonce="123">console.log("hi world")</script>
    HTML

    response2 = html_response(<<~HTML)
      <script nonce="456">console.log("hi world")</script>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it 'generates different ETags for the same content with and without a Rack session' do
    response1 = html_response('content', env: { 'rack.session' => session(id: '1') })
    response2 = html_response('content', env: {} )
    expect(response1).to_not have_same_etag_as(response2)
  end

  it 'generates different ETags for the same content with different Rack sessions' do
    response1 = html_response('content', env: { 'rack.session' => session(id: '1') })
    response2 = html_response('content', env: { 'rack.session' => session(id: '2') })
    expect(response1).to_not have_same_etag_as(response2)
  end

  it 'generates different ETags for the sameRack session when a Rails app manually rotates the CSRF token' do
    response1 = html_response('content', env: { 'rack.session' => session(id: '1', '_csrf_token': 'abc') })
    response2 = html_response('content', env: { 'rack.session' => session(id: '1', '_csrf_token': 'def') })
    expect(response1).to_not have_same_etag_as(response2)
  end

  it 'generates weak ETags because we only digest the response body' do
    response = html_response('Foo')
    expect(response[1]['ETag']).to start_with("W/")
  end

  # it 'does not crash with a Rack::BodyProxy' do
  #   app = lambda { |env| [200, { 'Content-Type' => 'text/html' }, Rack::BodyProxy.new("Hello, World!") {}] }
  #   response = etag(app).call(request)
  #   expect(response[1]['ETag']).to eq "W/\"dffd6021bb2bd5b0af676290809ec3a5\""
  # end

  it 'only strips patterns for HTML responses' do
    excel_content_type = { 'Content-Type' => 'application/vnd.ms-excel' }

    response1 = text_response(<<~HTML, headers: excel_content_type)
      <head>
        <meta name="csrf-token" content="6EueAlhls9P" />
      </head>
    HTML

    response2 = text_response(<<~HTML, headers: excel_content_type)
      <head>
        <meta name="csrf-token" content="qMN0fkVqOg" />
      </head>
    HTML

    expect(response1).not_to have_same_etag_as(response2)
  end

  it 'does strip pattern for a HTML content type with an appended charset' do
    html_content_type_with_charset = { 'Content-Type' => 'text/html; charset=utf-8' }

    response1 = text_response(<<~HTML, headers: html_content_type_with_charset)
      <head>
        <meta name="csrf-token" content="6EueAlhls9P" />
      </head>
    HTML

    response2 = text_response(<<~HTML, headers: html_content_type_with_charset)
      <head>
        <meta name="csrf-token" content="qMN0fkVqOg" />
      </head>
    HTML

    expect(response1).to have_same_etag_as(response2)
  end

  it 'generates ETags for non-HTML responses' do
    excel_content_type = { 'Content-Type' => 'application/vnd.ms-excel' }
    response1 = text_response('foo', headers: excel_content_type)
    response2 = text_response('foo', headers: excel_content_type)

    expect(response1).to have_same_etag_as(response2)
  end

  it "does not crash when generating ETags for binary responses that cannot be UTF-8 encoded" do
    excel_content_type = { 'Content-Type' => 'application/vnd.ms-excel' }
    not_utf8 = "vandflyver \xC5rhus"
    expect { text_response(not_utf8, headers: excel_content_type) }.not_to raise_error
  end

  # Tests from Rack::Test

  it "set ETag if none is set if status is 200" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html' }, ["Hello, World!"]] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to eq "W/\"dffd6021bb2bd5b0af676290809ec3a5\""
  end

  it "set ETag if none is set if status is 201" do
    app = lambda { |env| [201, { 'Content-Type' => 'text/html' }, ["Hello, World!"]] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to eq "W/\"dffd6021bb2bd5b0af676290809ec3a5\""
  end

  it "set Cache-Control to 'max-age=0, private, must-revalidate' (default) if none is set" do
    app = lambda { |env| [201, { 'Content-Type' => 'text/html' }, ["Hello, World!"]] }
    response = etag(app).call(request)
    expect(response[1]['Cache-Control']).to eq 'max-age=0, private, must-revalidate'
  end

  it "set Cache-Control to chosen one if none is set" do
    app = lambda { |env| [201, { 'Content-Type' => 'text/html' }, ["Hello, World!"]] }
    response = etag(app, nil, 'public').call(request)
    expect(response[1]['Cache-Control']).to eq 'public'
  end

  it "set a given Cache-Control even if digest could not be calculated" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html' }, []] }
    response = etag(app, 'no-cache').call(request)
    expect(response[1]['Cache-Control']).to eq 'no-cache'
  end

  it "sets a given Cache-Control for HTTP status codes that we don't digest, to preserve compatibility with Rack::ETag" do
    app = lambda { |env| [500, { 'Content-Type' => 'text/html' }, ["Hello, World!"]] }
    response = etag(app, 'no-store').call(request)
    expect(response[1]['Cache-Control']).to eq 'no-store'
  end

  it "not set Cache-Control if it is already set" do
    app = lambda { |env| [201, { 'Content-Type' => 'text/html', 'Cache-Control' => 'public' }, ["Hello, World!"]] }
    response = etag(app).call(request)
    expect(response[1]['Cache-Control']).to eq 'public'
  end

  it "not set Cache-Control if directive isn't present" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html' }, ["Hello, World!"]] }
    response = etag(app, nil, nil).call(request)
    expect(response[1]['Cache-Control']).to be_nil
  end

  it "not change ETag if it is already set" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html', 'ETag' => '"abc"' }, ["Hello, World!"]] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to eq "\"abc\""
  end

  it "not set ETag if body is empty" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html', 'Last-Modified' => Time.now.httpdate }, []] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to be_nil
  end

  it "not set ETag if Last-Modified is set" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html', 'Last-Modified' => Time.now.httpdate }, ["Hello, World!"]] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to be_nil
  end

  it "not set ETag if a sendfile_body is given" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html' }, sendfile_body] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to be_nil
  end

  it "not set ETag if a status is not 200 or 201" do
    app = lambda { |env| [401, { 'Content-Type' => 'text/html' }, ['Access denied.']] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to be_nil
  end

  it "set ETag even if no-cache is given" do
    app = lambda { |env| [200, { 'Content-Type' => 'text/html', 'Cache-Control' => 'no-cache, must-revalidate' }, ['Hello, World!']] }
    response = etag(app).call(request)
    expect(response[1]['ETag']).to eq "W/\"dffd6021bb2bd5b0af676290809ec3a5\""
  end

  it "close the original body" do
    body = StringIO.new
    app = lambda { |env| [200, { 'Content-Type' => 'application/octet-stream' }, body] }
    response = etag(app).call(request)
    expect(body).to_not be_closed
    response[2].close
    expect(body).to be_closed
  end

end


lib/rack/steady_etag.rb
require "rack"
require 'digest/sha2'
require_relative "steady_etag/version"

module Rack

  # Based on Rack::Etag from rack 2.2.3
  # https://github.com/rack/rack/blob/v2.2.3/lib/rack/etag.rb
  #
  # Automatically sets the ETag header on all String bodies.
  #
  # The ETag header is skipped if ETag or Last-Modified headers are sent or if
  # a sendfile body (body.responds_to :to_path) is given (since such cases
  # should be handled by apache/nginx).
  class SteadyETag

    # Yes, Rack::ETag sets a default Cache-Control for responses that it can digest.
    DEFAULT_CACHE_CONTROL = "max-age=0, private, must-revalidate"

    STRIP_PATTERNS = [
      /<meta\b[^>]*\bname=(["'])csrf-token\1[^>]+>/i,
      /<meta\b[^>]*\bname=(["'])csp-nonce\1[^>]+>/i,
      /<input\b[^>]*\bname=(["'])authenticity_token\1[^>]+>/i,
      lambda { |string| string.gsub(/(<script\b[^>]*)\bnonce=(["'])[^"']+\2+/i, '\1') }
    ]

    STRIP_CONTENT_TYPES = %w[
      text/html
      application/xhtml+xml
    ]

    def initialize(app, no_digest_cache_control = nil, digest_cache_control = DEFAULT_CACHE_CONTROL)
      @app = app

      @digest_cache_control = digest_cache_control

      # Rails sets a default `Cache-Control: no-cache` for responses that we cannot digest.
      # See https://github.com/rails/rails/blob/d96609505511a76c618dc3adfa3ca4679317d008/railties/lib/rails/application/default_middleware_stack.rb#L81
      @no_digest_cache_control = no_digest_cache_control
    end

    def call(env)
      status, headers, body = @app.call(env)
      headers = case_insensitive_headers(headers)
      session = env['rack.session']

      if etag_status?(status) && etag_body?(body) && !skip_caching?(headers)
        original_body = body
        digest, new_body = digest_body(body, headers, session)
        body = Rack::BodyProxy.new(new_body) do
          original_body.close if original_body.respond_to?(:close)
        end
        headers['ETag'] = %(W/"#{digest}") if digest
      end

      # It would make more sense to only set a Cache-Control for responses that we process.
      # However, the original Rack::ETag sets Cache-Control: @no_digest_cache_control
      # for all responses, even responses that we don't otherwise modify.
      # Hence if we move this code into the `if` above we would remove Rails' default
      # Cache-Control headers for non-digestable responses, which would be a considerable
      # change in behavior.
      if digest
        set_cache_control_with_digest(headers)
      else
        set_cache_control_without_digest(headers)
      end

      [status, headers, body]
    end

    private

    # HTTP headers are case-insensitive.
    # Wrap hedders into a hash with case-insensitive keys
    def case_insensitive_headers(headers)
      case Rack.release[0]
      when '1'
        Utils::HeaderHash.new(headers)
      when '2'
        Utils::HeaderHash[headers]
      when '3'
        raise "HeaderHash will be removed in Rack 3. Probably switch to new Headers."
      end
    end

    def set_cache_control_with_digest(headers)
      headers['Cache-Control'] ||= @digest_cache_control if @digest_cache_control
    end

    def set_cache_control_without_digest(headers)
      headers['Cache-Control'] ||= @no_digest_cache_control if @no_digest_cache_control
    end

    def etag_status?(status)
      status == 200 || status == 201
    end

    def etag_body?(body)
      # Rack main branch checks for `:to_ary` here to exclude streaming responses,
      # but that had other issues for me in testing. Maybe recheck when there is a
      # new Rack release after 2.2.3.
      !body.respond_to?(:to_path)
    end

    def skip_caching?(headers)
      headers.key?('ETag') || headers.key?('Last-Modified')
    end

    def digest_body(body, headers, session)
      parts = []
      digest = nil

      strippable_response = strippable_response?(headers)

      body.each do |part|
        parts << part

        # Note that `part` can be a string with binary data here.
        # It's important to check emptiness with #empty? instead of #blank?, since #blank?
        # internally calls String#match? and that explodes if the string is not valid UTF-8.
        unless part.empty?
          digest ||= initialize_digest(session)
          part = strip_patterns(part) if strippable_response
          digest << part
        end
      end

      if digest
        digest = digest.hexdigest.byteslice(0,32)
      end

      [digest, parts]
    end

    def initialize_digest(session)
      digest = Digest::SHA256.new

      if session
        if (session_id = session['session_id'])
          digest << session_id.to_s
        end

        # When we sign in or out with Devise, we always get a new session ID
        # and CSRF token. Lets anyway include the real (unmasked) CSRF token in the
        # digest in case a Rails controller manually rotates the token.
        if (rails_csrf_token = session['_csrf_token'])
          digest << rails_csrf_token.to_s
        end
      end

      digest
    end

    def strip_patterns(html)
      STRIP_PATTERNS.each do |pattern|
        if pattern.respond_to?(:call)
          html = pattern.call(html)
        else
          html = html.gsub(pattern, '')
        end
      end
      html
    end

    private

    def strippable_response?(headers)
      content_type = headers['Content-Type']
      return false unless content_type

      # Convert "text/tml; charset=utf-8" to just "text/html"
      content_type = content_type.split(/\s*;\s*/)[0]
      STRIP_CONTENT_TYPES.include?(content_type)
    end

  end
end
